File: typescript/src/tool.ts
/**
 * `defineUcTool` — declarative UC-syncable tool factory.
 *
 * TypeScript counterpart to Python's ``@tool(uc=..., grant=[...])`` decorator
 * (see ``python/src/apx_agent/_tool.py``). The Python version is a decorator
 * because Python tools are plain functions; here, tools are already
 * constructed via the {@link defineTool} factory in ``agent/tools.ts``, so we
 * follow the same factory pattern and add a complementary
 * {@link defineUcTool} for the UC-syncable case.
 *
 * Design choice — factory, not extension of `defineTool`:
 *
 *   We chose to ship `defineUcTool` as a separate factory rather than
 *   extending `defineTool` with optional `uc`/`grant` fields. Two reasons:
 *
 *     1. `defineTool` is provider-agnostic — it produces an `AgentTool` that
 *        any LLM provider can call. UC sync is Databricks-specific.
 *        Threading optional UC fields into the base factory leaks platform
 *        concerns into the generic tool shape.
 *     2. Validation of the three-part UC name should fail loudly at tool
 *        construction time. Making `uc` required on the dedicated factory
 *        keeps that contract explicit; making it optional on the base
 *        factory would dilute it.
 *
 *   `defineUcTool` wraps `defineTool` directly and attaches `ToolMetadata`
 *   plus a `uc_function` {@link ResourceSpec} to the returned `AgentTool`
 *   via the WeakMap registries in `resources.ts` and this file.
 *
 * Python semantics that did NOT translate:
 *
 *   - **Dependencies rejection.** Python rejects ``@tool(uc=...)`` on
 *     functions with ``Dependencies.*`` parameters because UC functions
 *     execute server-side under the function owner's identity and can't host
 *     a user-scoped `WorkspaceClient`. TypeScript has no equivalent
 *     `Dependencies` injection mechanism — tools are plain async closures
 *     that capture whatever they need lexically. There's nothing to inspect
 *     at construction time, so the check is omitted. If/when a TS
 *     `Dependencies`-style injection layer ships, this is the spot to add
 *     the same gate.
 *   - **`name` override mutates `__name__` / `__qualname__`.** Python rebinds
 *     the function's identity attributes. The TS equivalent — overriding the
 *     `AgentTool.name` field — happens naturally through the factory because
 *     the caller passes `name` directly to `defineTool`. The metadata
 *     records the override so callers that care can introspect it.
 */

import { z } from 'zod';

import { defineTool, type AgentTool } from './agent/tools.js';
import { attachResources, makeResourceSpec } from './resources.js';

// ---------------------------------------------------------------------------
// Metadata type
// ---------------------------------------------------------------------------

/**
 * Metadata attached to an {@link AgentTool} produced by {@link defineUcTool}.
 *
 * Stored in a module-private WeakMap keyed by the tool object — no
 * enumerable property is added to the returned tool. Read via
 * {@link getToolMetadata}.
 *
 * Frozen at construction.
 */
export interface ToolMetadata {
  /**
   * Fully-qualified UC function name (`catalog.schema.fn`). Always set on
   * tools produced by `defineUcTool`; the `| null` branch exists so future
   * Python-only-style tooling can share the same shape.
   */
  readonly ucName: string | null;
  /**
   * UC principals (users, groups, service principals) that should have
   * `EXECUTE` on the synced function. Empty when grants are managed
   * out-of-band or no UC sync is requested.
   */
  readonly grants: readonly string[];
  /**
   * Tool description supplied to the factory, mirrored here for
   * introspection. Set whenever `defineUcTool` is called (the description
   * is required on the underlying `defineTool`).
   */
  readonly descriptionOverride: string | null;
  /**
   * Tool name supplied to the factory, mirrored here. Always non-null on
   * tools produced by `defineUcTool`.
   */
  readonly nameOverride: string | null;
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

function validateUcName(uc: string): void {
  const segments = uc.split('.');
  if (segments.length !== 3) {
    throw new Error(
      `defineUcTool({ uc: ${JSON.stringify(uc)} }) must be a fully-qualified ` +
        `three-part UC name of the form catalog.schema.function. ` +
        `Got: ${JSON.stringify(uc)}`,
    );
  }
  for (const part of segments) {
    if (part.length === 0) {
      throw new Error(
        `defineUcTool({ uc: ${JSON.stringify(uc)} }) has an empty segment. ` +
          `Each of catalog, schema, and function must be a non-empty identifier.`,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// Metadata registry
// ---------------------------------------------------------------------------

/**
 * WeakMap from tool object → ToolMetadata. Separate from the resources
 * registry in `resources.ts` so the two concerns can evolve independently.
 */
const metadataRegistry = new WeakMap<object, ToolMetadata>();

function attachMetadata<T extends object>(tool: T, meta: ToolMetadata): T {
  metadataRegistry.set(tool, meta);
  return tool;
}

/**
 * Return the {@link ToolMetadata} attached to `tool` by {@link defineUcTool},
 * or `null` if none.
 */
export function getToolMetadata(tool: object | null | undefined): ToolMetadata | null {
  if (tool === null || tool === undefined) return null;
  return metadataRegistry.get(tool) ?? null;
}

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

/**
 * Options accepted by {@link defineUcTool}. Mirrors `defineTool` plus the
 * UC-sync fields. Run-time validation of `uc` (and the lightweight shape of
 * `grant`) happens through a Zod schema; the tool shape itself is checked
 * structurally by TypeScript.
 */
const ucOptionsSchema = z.object({
  uc: z.string().min(1, 'defineUcTool: uc is required'),
  grant: z.array(z.string().min(1)).optional(),
});

export interface DefineUcToolOptions<T extends z.ZodType> {
  /** Tool name shown to the LLM. */
  name: string;
  /** Tool description shown to the LLM. */
  description: string;
  /** Zod schema for tool arguments. */
  parameters: T;
  /** Async tool handler. Receives parsed arguments. */
  handler: (args: z.infer<T>) => Promise<unknown>;
  /**
   * Fully-qualified UC function name (`catalog.schema.function`) to sync
   * this tool into. Required. Validated as exactly three non-empty
   * dot-separated segments.
   */
  uc: string;
  /**
   * UC principals that should have `EXECUTE` on the synced function. Empty
   * by default.
   */
  grant?: string[];
}

/**
 * Define a UC-syncable agent tool.
 *
 * Returns the same {@link AgentTool} shape as {@link defineTool}, with
 * {@link ToolMetadata} attached (queryable via {@link getToolMetadata}) and
 * a `uc_function` {@link makeResourceSpec} attached via `attachResources`
 * so `collectResourceSpecs` picks it up automatically at compile/log time.
 *
 * @example
 *   const classifyIntent = defineUcTool({
 *     name: 'classify_intent',
 *     description: 'Classify a customer query as billing/technical/other.',
 *     parameters: z.object({ query: z.string() }),
 *     handler: async ({ query }) => (query.includes('bill') ? 'billing' : 'other'),
 *     uc: 'main.agents.tools.classify_intent',
 *     grant: ['agent_consumers'],
 *   });
 */
export function defineUcTool<T extends z.ZodType>(opts: DefineUcToolOptions<T>): AgentTool {
  // Validate the UC-specific options. Zod first (catches missing/empty uc),
  // then the structural three-part check.
  const parsed = ucOptionsSchema.parse({ uc: opts.uc, grant: opts.grant });
  validateUcName(parsed.uc);

  const grants: readonly string[] = Object.freeze([...(parsed.grant ?? [])]);

  const tool = defineTool({
    name: opts.name,
    description: opts.description,
    parameters: opts.parameters,
    handler: opts.handler,
  });

  const metadata: ToolMetadata = Object.freeze({
    ucName: parsed.uc,
    grants,
    descriptionOverride: opts.description,
    nameOverride: opts.name,
  });
  attachMetadata(tool, metadata);

  // Auto-declare the UC function as a Mosaic AI resource so log_agent /
  // collectResourceSpecs picks it up without explicit `extra` listing.
  attachResources(tool, [makeResourceSpec('uc_function', parsed.uc)]);

  return tool;
}

File: python/src/apx_agent/_tool.py
"""@tool decorator — declarative entry point for agent tools.

Two forms:

  * ``@tool`` — bare. Marks a function as an apx-agent tool with no changes
    to runtime behavior. Sugar for the typed-tool pattern that already works
    when you register a plain function via ``Agent(tools=[fn])``.

  * ``@tool(uc="catalog.schema.function", grant=["principal", ...])`` —
    marks the function for synchronisation to a Unity Catalog Python
    function. After ``publish_tools_to_uc(agent)``, the function exists as
    a governed UC asset reachable from Genie, Managed MCP, other agents,
    and any UC-aware client — without redefinition.

Design rules locked at the framework level:

  1. **UC-syncable iff pure.** ``uc=...`` is rejected at decoration time if
     the function has any ``Dependencies.*`` parameter. UC functions
     execute server-side under the function owner's identity, so a
     user-scoped ``WorkspaceClient`` is unavailable inside them. Tools
     needing the calling user's identity stay Python-only and don't sync.

  2. **Python is the source of execution.** At runtime, the tool runs as
     the Python function in-process. The UC function is the discovery /
     external-composition surface (Genie, MCP, sibling agents), not the
     execution path during the LLM loop. Avoids per-tool SQL roundtrips
     and keeps debugging local.

  3. **Explicit, fully-qualified UC names.** ``uc=...`` must be a three-part
     ``catalog.schema.function`` identifier. No implicit namespacing, no
     surprise writes. The decorator raises on shorter names.

  4. **Declarative grants.** ``grant=[...]`` declares the UC principals that
     should have ``EXECUTE`` on the function. ``publish_tools_to_uc``
     enforces this — code is the source of truth for who can call the
     tool.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Iterable, TypeVar, overload

from ._inspection import _inspect_tool_fn
from ._resources import ResourceSpec, attach_resources

_F = TypeVar("_F", bound=Callable[..., Any])


# ---------------------------------------------------------------------------
# Metadata type
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class ToolMetadata:
    """Metadata attached to a function by ``@tool``.

    Stored on the wrapped callable as ``_apx_tool``. Read by
    ``publish_tools_to_uc`` to decide which tools sync to Unity Catalog
    and what grants to apply.

    Attributes:
        uc_name: Fully-qualified UC function name (``catalog.schema.fn``)
            if the function should be synced to UC; ``None`` for
            Python-only tools.
        grants: UC principals (users, groups, service principals) that
            should have ``EXECUTE`` on the synced function. Empty when
            grants are managed out-of-band or no UC sync is requested.
        description_override: Tool description supplied to the decorator,
            overriding the function's ``__doc__``. ``None`` if absent.
        name_override: Tool name supplied to the decorator, overriding the
            Python function name. ``None`` if absent.
    """

    uc_name: str | None = None
    grants: tuple[str, ...] = field(default_factory=tuple)
    description_override: str | None = None
    name_override: str | None = None


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------


def _validate_uc_name(uc: str) -> None:
    if uc.count(".") != 2:
        raise ValueError(
            f"@tool(uc={uc!r}) must be a fully-qualified three-part UC name "
            f"of the form catalog.schema.function. Got: {uc!r}"
        )
    parts = uc.split(".")
    for part in parts:
        if not part:
            raise ValueError(
                f"@tool(uc={uc!r}) has an empty segment. Each of catalog, "
                f"schema, and function must be a non-empty identifier."
            )


def _validate_no_dependencies(fn: Callable[..., Any], uc: str) -> None:
    """Reject @tool(uc=...) on a function that uses Dependencies.*."""
    _, dep_param_names = _inspect_tool_fn(fn)
    if dep_param_names:
        raise ValueError(
            f"@tool(uc={uc!r}) cannot decorate {fn.__name__}: it has "
            f"Dependencies.* parameters {dep_param_names!r}. UC functions "
            f"execute server-side under the function owner's identity, so "
            f"user-scoped WorkspaceClient injection is unavailable. Either "
            f"remove the Dependencies parameter (and call any Databricks "
            f"APIs through the UC function's own identity) or drop uc=... "
            f"to keep the tool Python-only."
        )


# ---------------------------------------------------------------------------
# Decorator
# ---------------------------------------------------------------------------


@overload
def tool(fn: _F) -> _F: ...


@overload
def tool(
    *,
    name: str | None = ...,
    description: str | None = ...,
    uc: str | None = ...,
    grant: Iterable[str] | None = ...,
) -> Callable[[_F], _F]: ...


def tool(
    fn: _F | None = None,
    *,
    name: str | None = None,
    description: str | None = None,
    uc: str | None = None,
    grant: Iterable[str] | None = None,
) -> Any:
    """Mark a function as an apx-agent tool.

    Plain form — no args, equivalent to passing the function to
    ``Agent(tools=[...])`` directly::

        @tool
        def classify_intent(query: str) -> str:
            \"\"\"Classify a customer query as billing/technical/account/other.\"\"\"
            ...

    UC-syncable form — registers the function in Unity Catalog at deploy
    time so Genie, Managed MCP, and other agents can reach it through the
    catalog::

        @tool(uc="main.agents.tools.classify_intent", grant=["agent_consumers"])
        def classify_intent(query: str) -> str:
            \"\"\"Classify a customer query as billing/technical/account/other.\"\"\"
            ...

    Args:
        fn: The function being decorated (bare form).
        name: Override the tool name (shown to the LLM). Defaults to
            ``fn.__name__``.
        description: Override the tool description (shown to the LLM).
            Defaults to ``fn.__doc__``.
        uc: Fully-qualified UC function name (``catalog.schema.function``)
            to sync this tool into. When set, the tool is marked for
            registration by ``publish_tools_to_uc`` and contributes a
            ``DatabricksFunction`` resource at log time. Must be omitted
            for tools that use ``Dependencies.*`` parameters.
        grant: UC principals that should have ``EXECUTE`` on the synced
            function. Only meaningful when ``uc`` is set. Empty by default.

    Returns:
        The decorated function with ``_apx_tool`` metadata attached. Pass it
        to ``Agent(tools=[...])`` like any other tool.
    """

    grants_tuple = tuple(grant) if grant else ()

    def _decorate(target: _F) -> _F:
        if not callable(target):
            raise TypeError(f"@tool requires a callable; got {type(target).__name__}")

        if uc is not None:
            _validate_uc_name(uc)
            _validate_no_dependencies(target, uc)
        elif grants_tuple:
            raise ValueError(
                "@tool(grant=...) requires uc=... — grants only apply when "
                "the tool is synced to a UC function."
            )

        if name:
            target.__name__ = name
            target.__qualname__ = name
        if description:
            target.__doc__ = description

        metadata = ToolMetadata(
            uc_name=uc,
            grants=grants_tuple,
            description_override=description,
            name_override=name,
        )
        target._apx_tool = metadata  # type: ignore[attr-defined]

        # When UC-synced, the function becomes a declared resource so
        # log_agent picks it up automatically.
        if uc is not None:
            attach_resources(target, [ResourceSpec("uc_function", uc)])

        return target

    if fn is not None:
        # Bare @tool form
        return _decorate(fn)

    # @tool(...) form
    return _decorate


def get_tool_metadata(fn: Any) -> ToolMetadata | None:
    """Return ``ToolMetadata`` attached to ``fn`` by ``@tool``, or ``None``."""
    meta = getattr(fn, "_apx_tool", None)
    if isinstance(meta, ToolMetadata):
        return meta
    return None
